### DISCUSSION TITLE
External gNB

### DISCUSSION THREAD

**Discussion by abouillot-stl** [herlesupreeth/docker_open5gs#154] [Q&A] [answered]
Hi,

I succeeded having the Open5GS running in containers with the UERANSIM gNB and UE.

I want to move forward to physical world and use Amarisoft gNB in order to have real UEs connected.

I performed the suggested changes in the Configuration section of the readme.md here, by uncommenting  the various ports in the various yaml files, as well as setting the physical PC IP addresses in the .env file.

I noticed on https://www.sharetechnote.com/html/OpenRAN/OR_open5gs_Run.html which relies on a monolithic Open5GS VM setup they advertise the IP address in the `afm.yaml` file. Shouldn't and `AMF_ADVERTISE_IP` be added in the `.env` file and 
```
    ngap:
      - addr: AMF_IP
        advertise: AMF_ADVERTISE_IP
```
be set in `afm.yaml`, alongside with 
```
sed -i 's|AMF_ADVERTISE_IP|'$AMF_ADVERTISE_IP'|g' install/etc/open5gs/amf.yaml
```
in the `amf/amf_init.sh` files?

I was preparing a set of changes to push, but I probably miss something as the INIT message from gNB cannot reach its destination, according to Wireshark capture on the Open5GS end.

**Comment by herlesupreeth** [comment]
You dont need AMF advertise address since the docker is publishing port 38413 for NGAP communications over SCTP. 

Just make sure in sa-deploy.yaml file you uncomment ports section under AMF and UPF section. Then use it as follows

docker-compose - f sa-deploy.yaml up

**Comment by abouillot-stl** [comment]
So I guess my issue is somewhere else on the host/docker setup, as the external gNB or eNB are not able to establish communication with AMF or MME respectively.

**Reply by herlesupreeth** [reply]
Please post a wireshark trace taken when you attempt to connect gNB to AMF

**Reply by abouillot-stl** [reply]
Please find attached the pcap of the startup. There are three connection attempts, all failed.
![Screenshot from 2023-04-11 10-37-58](https://user-images.githubusercontent.com/90634041/231105709-d3fac074-85f4-46ab-936b-a6ef998eeed3.png)
[sa.zip](https://github.com/herlesupreeth/docker_open5gs/files/11198467/sa.zip)
Here the capture from the gNB side:
[gnb.zip](https://github.com/herlesupreeth/docker_open5gs/files/11198736/gnb.zip)

The setup is

TEST_NETWORK=172.22.0.0/24
DOCKER_HOST_IP=192.168.100.33


gNB IP=192.168.100.1

I also attach the diif with the out of github setup.

[sa_diff.txt](https://github.com/herlesupreeth/docker_open5gs/files/11198466/sa_diff.txt)

**Reply by herlesupreeth** [reply]
I suspect it could be some firewall on the host machine running docker_open5gs. Can you try running below command once and then giving it a try?

`sudo ufw disable`

**Reply by abouillot-stl** [reply]
Unfortunately this doesn't change the outcome. 

To be noted, the Open5GS web interface (port 3000) can be reached from the gNB PC when the FW is disabled, but not when enabled.

So one tiny step forward.

I setup a separate PC with UERANSIM on it. Set the host docker IP address as `amfConfigs/address` in the `custom-gnb.yaml`, and the PC physical IP address for the `linkIp`, `ngapIp` and `gtpIp` addresses. I observe the very same behavior, with a connection refused. That PC has access to the web UI.

**Reply by herlesupreeth** [reply]
Can you check whether your router is filtering out SCTP packets?

**Reply by abouillot-stl** [reply]
I can have exchange both ways between the two PC using `sctp_darn` test program.

**Reply by abouillot-stl** [reply]
On the gNB side (UERANSIM) should the amfConfigs address be 172.22.0.10 (i.e. .env container one) or 192.168.100.31 (i.e. Docker host physical address)?

With the former I only get SCTP INIT messages. With the later, I get SCTP INIT/ABORD packets on Open5GS side.

172.22.0.10 being possible with the help of route added on the gNB side via Docker physical address (`sudo ip route add 172.22.0.0/24 via 192.168.100.31`)

**Comment by herlesupreeth** [comment]
> I setup a separate PC with UERANSIM on it. Set the host docker IP address as amfConfigs/address in the custom-gnb.yaml, and the PC physical IP address for the linkIp, ngapIp and gtpIp addresses. I observe the very same behavior, with a connection refused. That PC has access to the web UI.

I realized later you would face this issue if you run UERANSIM in docker on another machine. If you run UERANSIM natively (no docker involved) on another PC then you would not face this issue of SCTP.

Are you running the gNB in a docker by any chance?

> On the gNB side (UERANSIM) should the amfConfigs address be 172.22.0.10 (i.e. .env container one) or 192.168.100.31 (i.e. Docker host physical address)?

It should be 192.168.100.31

**Reply by abouillot-stl** [reply]
UERANSIM is working natively on the separate PC. Thanks for hinting the docker potential issue, as I could have tried this later down the road. I'll keep away for time being.

I tried both IP option, but as I said, with similar unsuccess. I'll stick to 192.168.100.31.

**Reply by herlesupreeth** [reply]
Can you paste here the docker version and docker-compose version you are using? Just make sure there are no double install of docker. I have seen crazy issues when docker is installed manually while the docker came pre-installed via snap

**Reply by abouillot-stl** [reply]
I installed latest Docker Desktop, which came with docker compose v2:
```
user@open5gs-1:~$ sudo apt list --installed | grep docker
docker-buildx-plugin/focal,now 0.10.4-1~ubuntu.20.04~focal amd64 [installed,automatic]
docker-ce-cli/focal,now 5:23.0.3-1~ubuntu.20.04~focal amd64 [installed,automatic]
docker-compose-plugin/focal,now 2.17.2-1~ubuntu.20.04~focal amd64 [installed,automatic]
docker-desktop/now 4.18.0-104112 amd64 [installed,local]
```
Here is the output of `docker info` command:
[docker_info.txt](https://github.com/herlesupreeth/docker_open5gs/files/11209089/docker_info.txt)

When installing, I used `docker compose` (v2) in place of `docker-compose` (v1)

**Reply by herlesupreeth** [reply]
I am not sure whether it helps or not. But try updating your docker, its a bit old. You are running version 20.10.24 and the latest is 23.0.3. Older docker versions had flaky SCTP support.

**Reply by abouillot-stl** [reply]
I installed  the latest Docker Desktop, released 2023-04-03, as availlable on [www.docker.com](https://www.docker.com).

I uninstalled it and setup Docker Engine.
```
user@open5gs-1:~$ sudo apt list --installed | grep docker
[sudo] password for user: 
docker-buildx-plugin/focal,now 0.10.4-1~ubuntu.20.04~focal amd64 [installed]
docker-ce-cli/focal,now 5:23.0.3-1~ubuntu.20.04~focal amd64 [installed]
docker-ce-rootless-extras/focal,now 5:23.0.3-1~ubuntu.20.04~focal amd64 [installed,automatic]
docker-ce/focal,now 5:23.0.3-1~ubuntu.20.04~focal amd64 [installed]
docker-compose-plugin/focal,now 2.17.2-1~ubuntu.20.04~focal amd64 [installed]
```

Sole visible change is the exposition of internal docker network.

I took pcap of both physical link, as well as the bridge traffic. I don't see SCTP traffic with AMF component.

[docker_bridge.zip](https://github.com/herlesupreeth/docker_open5gs/files/11210119/docker_bridge.zip)
[host.zip](https://github.com/herlesupreeth/docker_open5gs/files/11210121/host.zip)

**Reply by herlesupreeth** [reply]
I see SCTP traffic on your host but not in trace take on docker bridge, this leads to my assumption that docker-compose file you are using does not have port 38412 port published under AMF section (not uncommented I mean)

**Reply by abouillot-stl** [reply, marked as answer]
What an eagle eye! On the retry (with Docker Engine, rather than Docker Desktop) I indeed started the container with `nsa-deploy.yaml` config.

When started with `sa-deploy.yaml`it works much better.

I checked and can confirm I used the proper `sa-deploy.yaml` when trying with Docker Desktop. I'll dig on that side for curiosity why it can't properly work in this setup.

**Comment by abouillot-stl** [comment]
Ok, After reseting the PC with Docker Desktop, I confirm the external eNB/gNB cannot reach the Core. When switching to Docker Engine - Exact same configuration files, all work as expected. Maybe an note in the readme could be added?

Also to note with latest version of Docker Engine comes Compose v2 in lieu of docker-compose. This will be removed in a couple of month. Maybe an update of the same readme?

**Comment by herlesupreeth** [comment]
> Maybe an note in the readme could be added?

I dont mind adding but the README states to use docker-ce and not docker desktop :)

> Also to note with latest version of Docker Engine comes Compose v2 in lieu of docker-compose. This will be removed in a couple of month. Maybe an update of the same readme?

Yeah, thats still on my TODO list. Its a matter of replacing docker-compose with docker compose so hadn't given much importance
